Services/ProtocolServer/GeminiDownload.h

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright (c) 2020, The SerenityOS developers.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include <AK/Badge.h>
+#include <LibCore/Forward.h>
+#include <LibGemini/GeminiJob.h>
+#include <ProtocolServer/Download.h>
+
+class GeminiProtocol;
+
+class GeminiDownload final : public Download {
+public:
+    virtual ~GeminiDownload() override;
+    static NonnullRefPtr<GeminiDownload> create_with_job(Badge<GeminiProtocol>, PSClientConnection&, NonnullRefPtr<Gemini::GeminiJob>&&);
+
+private:
+    explicit GeminiDownload(PSClientConnection&, NonnullRefPtr<Gemini::GeminiJob>&&);
```

Comment:
How does `&&`ing a `NonnullRefPtr` work, even?

Reply:
just forces it to be moved, I imagine?
IIRC `NonnullRefPtr` doesn't increment the refcount on move with `NonnullRefPtr&&`

Reply:
Yes, but how does moving a <code><strong>Nonnull</strong>RefPtr</code> work? What does it leave in place when you move out of it?

Reply:
The same content a before?
The impl seems to just leak the reference

Reply:
So it does leave null in place (not so non-null after all!), but at least it asserts if you try to access it later:

https://github.com/SerenityOS/serenity/blob/3c3c1b40131b41454c59c4f5c8f6eaf9094556a1/AK/NonnullRefPtr.h#L80-L84
https://github.com/SerenityOS/serenity/blob/3c3c1b40131b41454c59c4f5c8f6eaf9094556a1/AK/NonnullRefPtr.h#L165-L171

Reply:
Yeah, that's a bit surprising.
But I can see some sense in the behaviour, I guess